websockets/exceptions.py

```diff
@@ -26,13 +26,31 @@ class InvalidOrigin(InvalidHandshake):
     """
 
 
+class InvalidStatus(InvalidHandshake):
+    """
+    Exception raised when a handshake response status code is invalid.
+
+    Provides the integer status code and reason in its ``code`` and
+    ``reason`` attributes, respectively.
+
+    """
+    def __init__(self, code, reason=None):
+        self.code = code
+        self.reason = reason
+        message = 'Bad status code: {}'.format(code)
+        if reason:
+            message += ' ({})'.format(reason)
+        super().__init__(message)
+
+
 class InvalidState(Exception):
     """
     Exception raised when an operation is forbidden in the current state.
 
     """
 
 
+# TODO: DRY up with InvalidStatus?
```

Comment:
Actually, looking back at this, the other thought I had in mind (in addition to the possibility of subclassing) was to DRY up the code responsible for constructing the message attribute (in which case the more appropriate location for the TODO would be in the method body). This was for aesthetic reasons to ensure the message format would be identical for the user, and for normal code reasons. In InvalidStatus, I used the format: `f"{description}: {code} ({reason})"`.

Whether or not you agree with the change (which could be done in a separate PR), I can remove the TODO. It was mainly to spark the conversation we are currently having as to whether code should be shared. :)